File: src/backend/app/api/routes/claims.py
from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.auth.models import User
from app.database import get_database_session
from app.items.service import DomainError
from app.multimodal.factory import get_multimodal_adapter
from app.multimodal.ports import MultimodalPort
from app.settings import settings
from app.reviews.schemas import ReviewRequestCreate
from app.reviews.service import create_claim_review_request
from app.verification.schemas import (
    ClaimOutcome,
    IdentityClaimRequest,
    OtherClaimRequest,
    QuestionPublic,
)
from app.verification.service import (
    get_other_questions,
    submit_identity_claim,
    submit_other_claim,
)


router = APIRouter(prefix="/api/candidates", tags=["claims"])
claim_review_router = APIRouter(prefix="/api/claims", tags=["claims"])


@router.post("/{candidate_id}/claims/identity", response_model=ClaimOutcome)
async def identity_claim(
    candidate_id: UUID,
    payload: IdentityClaimRequest,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_database_session),
) -> ClaimOutcome:
    try:
        result = await submit_identity_claim(
            session,
            candidate_id=candidate_id,
            requester_id=user.id,
            full_number=payload.full_number,
            hmac_key=settings.id_hmac_key_v1.encode("utf-8"),
        )
        await session.commit()
        return result
    except DomainError:
        await session.rollback()
        raise


@router.get("/{candidate_id}/questions", response_model=list[QuestionPublic])
async def other_questions(
    candidate_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_database_session),
) -> list[QuestionPublic]:
    return await get_other_questions(
        session, candidate_id=candidate_id, requester_id=user.id
    )


@router.post("/{candidate_id}/claims/answers", response_model=ClaimOutcome)
async def other_claim(
    candidate_id: UUID,
    payload: OtherClaimRequest,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_database_session),
    adapter: MultimodalPort = Depends(get_multimodal_adapter),
) -> ClaimOutcome:
    try:
        result = await submit_other_claim(
            session,
            candidate_id=candidate_id,
            requester_id=user.id,
            answers={answer.question_id: answer.answer for answer in payload.answers},
            adapter=adapter,
        )
        await session.commit()
        return result
    except DomainError:
        await session.rollback()
        raise


@claim_review_router.post("/{claim_id}/review-requests")
async def claim_review_request(
    claim_id: UUID,
    payload: ReviewRequestCreate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_database_session),
) -> dict[str, str]:
    try:
        request = await create_claim_review_request(
            session,
            claim_id=claim_id,
            requester_id=user.id,
            reason=payload.reason,
        )
        await session.commit()
    except DomainError:
        await session.rollback()
        raise
    return {"id": str(request.id), "status": request.status}

File: frontend/src/api/claims.ts
import { apiClient, isMockMode } from './client'
import { mockApi } from './mock'
import type { ClaimDetail, ClaimOutcome, QuestionPublic } from './types'

export const claimsApi = {
  async get(claimId: string): Promise<ClaimDetail> {
    if (isMockMode) return mockApi.unsupported('认领详情')
    return (await apiClient.get<ClaimDetail>(`/api/claims/${claimId}`)).data
  },

  async questions(candidateId: string): Promise<QuestionPublic[]> {
    if (isMockMode) return mockApi.unsupported('认领问题')
    return (await apiClient.get<QuestionPublic[]>(`/api/candidates/${candidateId}/questions`)).data
  },

  async verifyIdentity(candidateId: string, fullNumber: string): Promise<ClaimOutcome> {
    if (isMockMode) return mockApi.unsupported('证件认领核验')
    return (await apiClient.post<ClaimOutcome>(`/api/candidates/${candidateId}/claims/identity`, { full_number: fullNumber })).data
  },

  async verifyAnswers(candidateId: string, answers: Array<{ question_id: string; answer: string }>): Promise<ClaimOutcome> {
    if (isMockMode) return mockApi.unsupported('问题认领核验')
    return (await apiClient.post<ClaimOutcome>(`/api/candidates/${candidateId}/claims/answers`, { answers })).data
  },

  async createReview(claimId: string, reason: string): Promise<{ id: string; status: string }> {
    if (isMockMode) return mockApi.unsupported('认领复核')
    return (await apiClient.post<{ id: string; status: string }>(`/api/claims/${claimId}/review-requests`, { reason })).data
  },

  async contact(claimId: string): Promise<{ finder_phone: string; finder_email: string }> {
    if (isMockMode) return mockApi.unsupported('交接联系方式')
    return (await apiClient.get<{ finder_phone: string; finder_email: string }>(`/api/claims/${claimId}/contact`)).data
  },

  async completeHandoff(claimId: string, idempotencyKey: string): Promise<{ claim_id: string; status: string }> {
    if (isMockMode) return mockApi.unsupported('完成交接')
    return (await apiClient.post<{ claim_id: string; status: string }>(`/api/claims/${claimId}/handoff-complete`, {
      confirmation: true,
    }, { headers: { 'Idempotency-Key': idempotencyKey } })).data
  },
}
